main.cpp

```diff
@@ -0,0 +1,201 @@
+#include <iostream>
+#include <stdbool.h>
+#include <vector>
+#include <tuple>
+
+#include <climits>
+
+#include <cmath>
+#include <cassert>
+
+typedef int dtype; 
+
+class Square {
+private:
+    dtype num; 
+    //List of Possibilities
+
+public:
+    Square(): num(-1) {}
+    Square(dtype val): num(val) {}
+
+    bool is_solved() { return (num < 0) ? false : true; };
+    dtype& value() {
+        return num;
+    }
+    
+    //override = operator, so can assign value directly
+    void operator= (const dtype num) {
+        this->num = num;
+    }
+};
+
+class Grid {
+public:
+    class GridRow
+    {
+        private:
+            friend class Grid; // Allow Grid to Create & Access Internals
+            std::vector<Square> vec;
+            GridRow(const size_t size) : vec(size, Square()){}
+            GridRow(const std::vector<dtype> v) {
+                for(auto s: v) {
+                    vec.emplace_back(Square(s));
+                }
+            } 
+            
+        public:
+            Square& operator[](size_t idx) {return vec[idx];}
+            size_t size() {return vec.size();}
+            std::vector<dtype> values() {
+                std::vector<dtype> res;
+                for(auto sq: vec) {
+                    res.emplace_back(sq.value());
+                }
+                return res;
+            }
+    };
+    
+private:
+    size_t grid_size; //row & col 
+    size_t block_len; // sqrt(row)
+
+    std::vector<GridRow> grid;
+    
+    //checks is the vectors are possible in sudoku
+    //returns : bool - true or false if possible 
+    //   size_t - number of digits missing in row (only valid when bool is true)
+    std::pair<bool, size_t> check_if_vecs_possible(const 

```

Comment:
Thank you for your great effort. However, I feel like this function needs to frequently be called when doing search and at the end validating. Each time, you load in an entire vector to check if it is possible, which can be sort of expensive. I don't know if you have seen my email about https://github.com/Hao2747/sudoku-online/blob/master/src/solver_basic.cc code. Their idea is we maintain some states, keeping check of possible availabilities for each row/ colomn/block. Each time you want to add one number, you just need to check if that number is used already. How do you think? we can also setup a zoom meeting to discuss about this.